File: MenstralWahala/MenstralWahalajs/MenstralApp.js
module.exports = {tracker, expectingNextPeriod, checkRegularOrNot, fertileWindow}

const tracker = function(currentPeriod, cycleLength) {
   
    const [day, month, year] = currentPeriod.split('-').map(Number);
    
    const currentDate = new Date(year, month - 1, day); 
    
    currentDate.setDate(currentDate.getDate() + cycleLength);
    
    const nextDay = String(currentDate.getDate()).padStart(2, '0');
    const nextMonth = String(currentDate.getMonth() + 1).padStart(2, '0'); // month is 0-indexed
    const nextYear = currentDate.getFullYear();

    return `${nextDay}-${nextMonth}-${nextYear}`;
}

constt fertileWindow = function(currentPeriod, cycleLength) {
    const formatter = (date) => {
        const day = String(date.getDate()).padStart(2, '0');
        const month = String(date.getMonth() + 1).padStart(2, '0');
        const year = date.getFullYear();
        return `${day}-${month}-${year}`;
    };

    const trackerDate = tracker(currentPeriod, cycleLength);
    const [tDay, tMonth, tYear] = trackerDate.split('-').map(Number);
    const periodDate = new Date(tYear, tMonth - 1, tDay);

    const ovulationDate = new Date(periodDate);
    ovulationDate.setDate(ovulationDate.getDate() - 14);

    const fertileStart = new Date(ovulationDate);
    fertileStart.setDate(fertileStart.getDate() - 5);

    const fertileEnd = new Date(ovulationDate);
    fertileEnd.setDate(fertileEnd.getDate() + 1);

    return (
        `Ovulation Date: ${formatter(ovulationDate)}\n` +
        `Fertile Window: ${formatter(fertileStart)} to ${formatter(fertileEnd)}`
    );
}
return `Ovulation Date: ${formatDate(ovulationDate)}\nFertile Window: ${formatDate(fertileStart)} to ${formatDate(fertileEnd)}`;
}


const checkRegularOrNot = function(periodDates, cycleLength) {
    const formatter = (dateStr) => {
        const [day, month, year] = dateStr.split('-').map(Number);
        return new Date(year, month - 1, day);
    };

    const actualCycleLengths = [];

    for (let i = 1; i < periodDates.length; i++) {
        const previous = formatter(periodDates[i - 1]);
        const current = formatter(periodDates[i]);
        const diffInDays = (current - previous) / (1000 * 60 * 60 * 24); // convert ms to days
        actualCycleLengths.push(diffInDays);
    }

    for (const cycle of actualCycleLengths) {
        if (Math.abs(cycle - cycleLength) > 2) {
            return "Irregular";
        }
    }

    return "Regular";
}

File: MenstralWahala/MenstralWahalapy/menstral_app.py
from datetime import datetime, timedelta

def tracker(current_period: str, cycle_length: int): 
    formatter = "%d-%m-%Y"
    current_date = datetime.strptime(current_period, formatter)
    next_period_date = current_date + timedelta(days=cycle_length)
    return next_period_date.strftime(formatter)

def fertile_window(current_period: str, cycle_length: int): 
    formatter = "%d-%m-%Y"
    the_tracker = tracker(current_period, cycle_length)
    period_date = datetime.strptime(the_tracker, formatter)
    ovulation_date = period_date - timedelta(days=14)
    fertile_start = ovulation_date - timedelta(days=5)
    fertile_end = ovulation_date + timedelta(days=1)

    return (
        f"Ovulation Date: {ovulation_date.strftime(formatter)}\n"
        f"Fertile Window: {fertile_start.strftime(formatter)} to {fertile_end.strftime(formatter)}"
    )

def check_regular_or_not(period_dates: list[str], cycle_length: int): 
    formatter = "%d-%m-%Y"
    actual_cycle_lengths = []

    for count in range(1, len(period_dates)):
        previous = datetime.strptime(period_dates[count - 1], formatter)
        current = datetime.strptime(period_dates[count], formatter)
        cycle = (current - previous).days
        actual_cycle_lengths.append(cycle)

    for cycle in actual_cycle_lengths:
        if abs(cycle - cycle_length) > 2:
            return "Irregular"
    
    return "Regular"


